File: MirrorMirror/MirrorMirror/Controllers/SLAnimatedViewControllerContainer.m
//
//  SLAnimatedViewControllerContainer.m
//  MirrorMirror
//
//  Created by Bradley Griffith on 6/1/14.
//  Copyright (c) 2014 Summer League. All rights reserved.
//

#import "SLAnimatedViewControllerContainer.h"

@interface SLAnimatedViewControllerContainer ()

@property (nonatomic, strong) UIViewController *rootViewController;

@end

@implementation SLAnimatedViewControllerContainer

- (id)initWithRootViewController:(UIViewController *)rootViewController
{
	self = [super init];
	if (self) {
		self.rootViewController = rootViewController;
	}
	return self;
}

- (void)addController:(UIViewController *)viewController
{
	[viewController willMoveToParentViewController:self];
	[self addChildViewController:viewController];
}

- (UIViewController *)topMostViewController
{
	return [self.childViewControllers lastObject];
}

- (UIViewController *)previousViewController
{
	if ([self.childViewControllers count] == 1) {
		return nil;
	}
	int index = [self.childViewControllers count] - 2;
	return self.childViewControllers[index];
}


// Note:
//   Subclassers should define how the both the currently displayed view controller, and the view controller to be displayed,
//   should begin and end their respective animations using the animationStart and animationEnd blocks.
- (void)pushViewController:(UIViewController *)viewController withDuration:(NSTimeInterval)duration animatioStart:(void (^)(void))animationStart animationEnd:(void (^)(void))animationEnd completion:(void (^)(void))completion
{
	[self addController:viewController];
	viewController.view.transform = CGAffineTransformIdentity;
	[self.view addSubview:viewController.view];
	
	animationStart();
	
	[UIView animateWithDuration:duration animations:^{
		animationEnd();
	} completion:^(BOOL finished) {
		[viewController didMoveToParentViewController:self];
		if (completion) {
			completion();
		}
	}];
}

// Note:
//   Subclassers should define how the both the currently displayed view controller, and the view controller to be displayed,
//   should begin and end their respective animations using the animationStart and animationEnd blocks.
- (void)popViewControllerWithDuration:(NSTimeInterval)duration animatioStart:(void (^)(void))animationStart animationEnd:(void (^)(void))animationEnd completion:(void (^)(void))completion
{
	if ([self.childViewControllers count] == 1) {
		return;
	}
	
	UIViewController *topVC = [self topMostViewController];
	
	[topVC willMoveToParentViewController:nil];
	
	UIViewController *targetVC = [self previousViewController];
	CGAffineTransform previousTransform = targetVC.view.transform;
	targetVC.view.transform = CGAffineTransformIdentity;
	targetVC.view.frame = self.view.bounds;
	targetVC.view.transform = previousTransform;
	
	animationStart();
	
	[UIView animateWithDuration:duration animations:^{
		animationEnd();
	} completion:^(BOOL finished) {
		[topVC.view removeFromSuperview];
		[topVC removeFromParentViewController];
		if (completion) {
			completion();
		}
	}];
}

- (void)viewDidLoad
{
	[super viewDidLoad];
	
	self.view.autoresizesSubviews = YES;
	
	if ([self.childViewControllers count] == 0) {
		[self addController:self.rootViewController];
		UIView *rootView = self.rootViewController.view;
		rootView.frame = self.view.bounds;
		[self.view addSubview:rootView];
		[self.rootViewController didMoveToParentViewController:self];
	}
}

@end


@implementation UIViewController (AnimatedContainer)

- (SLAnimatedViewControllerContainer *)animatedContainer {
	return (SLAnimatedViewControllerContainer *)self.parentViewController;
}

@end

File: MirrorMirror/MirrorMirror/Controllers/SLAnimatedViewControllerContainer.swift
//
//  SLAnimatedViewController.swift
//  MirrorMirror
//
//  Created by Bradley Griffith on 6/4/14.
//  Copyright (c) 2014 Summer League. All rights reserved.
//

import UIKit

class SLAnimatedViewControllerContainer: UIViewController {
	
	var rootViewController: UIViewController!
	
	
	/* Lifecycle
	------------------------------------------*/
	
	init(rootViewController: UIViewController) {
		
		self.rootViewController = rootViewController
		super.init(coder: nil)
	}
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		self.view.autoresizesSubviews = true
		
		if self.childViewControllers.count == 0 {
			self.addViewController(self.rootViewController)
			var rootView: UIView = self.rootViewController.view
			rootView.frame = self.view.bounds
			self.view.addSubview(rootView)
			self.rootViewController.didMoveToParentViewController(self)
		}
	}
	
	
	/* Instance Methods
	------------------------------------------*/
	
	func addViewController(viewController: UIViewController) {
		viewController.willMoveToParentViewController(self)
		self.addChildViewController(viewController)
		//self.childViewControllers
	}
	
	func topMostViewController() -> UIViewController {
		return self.childViewControllers[self.childViewControllers.endIndex - 1] as UIViewController
	}
	
	func previousViewController() -> UIViewController? {
		if self.childViewControllers.count == 1 {
			return nil
		}
		
		var index = self.childViewControllers.count - 2
		return self.childViewControllers[index] as? UIViewController
	}
	
	func pushViewController(viewController: UIViewController, duration: NSTimeInterval, animationStart: (() -> Void), animationEnd: (() -> Void), completion: (() -> Void)) {
		self.addViewController(viewController)
		viewController.view.transform = CGAffineTransformIdentity
		
		self.view.addSubview(viewController.view)
		
		animationStart()
		
		UIView.animateWithDuration(duration, animations: {
				animationEnd()
			}, completion: {
				_ in
				viewController.didMoveToParentViewController(self)
				
				completion()
			})
		
	}
	
	func popViewControllerWithDuration(duration: NSTimeInterval, animationStart: (() -> Void), animationEnd: (() -> Void), completion: (() -> Void)) {
		
		if self.childViewControllers.count == 1 {
			return
		}
		
		var topVC: UIViewController = self.topMostViewController()
		
		topVC.willMoveToParentViewController(nil)
		
		var targetVC: UIViewController = self.previousViewController()!
		var previousTransform: CGAffineTransform = targetVC.view.transform
		targetVC.view.transform = CGAffineTransformIdentity
		targetVC.view.frame = self.view.bounds
		targetVC.view.transform = previousTransform
		
		animationStart()
		
		UIView.animateWithDuration(duration, animations: {
				animationEnd()
			}, completion: {
				_ in
				topVC.view.removeFromSuperview()
				topVC.removeFromParentViewController()
				
				completion()
			})
	}
	
}
